File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        l1,l2=headA,headB
        if not l1 or not l2:
            return None
        char=set()
        while l1:
            if l1 in char:
                continue
            char.add(l1)
            l1=l1.next
        while l2:
            if l2 in char:
                return l2
            l2=l2.next
        return None

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        Set <ListNode> nodeMap = new HashSet <>();
        ListNode l1 = headA, l2 = headB;
        
        while (l1 != null){
            nodeMap.add(l1);
            l1 = l1.next;
        }
        while (l2 != null){
            if (nodeMap.contains(l2)) return l2;
            l2 = l2.next;
        }
        return null;   
    }
}
